uspace/srv/hid/input/input.c

```diff
@@ -107,6 +109,61 @@ static list_t serial_devs;
 
 static FIBRIL_MUTEX_INITIALIZE(discovery_lock);
 
+/* changes all kb devices to the given layout */
+static void layout_change(layout_ops_t *layout)
+{
+	list_foreach(kbd_devs, link, kbd_dev_t, kdev) {
+		layout_t *ret = layout_create(layout);
+		if (ret != NULL) {
+			layout_destroy(kdev->active_layout);
+			kdev->active_layout = ret;
+		}
+	}
+}
+
+/* similiar like layout_change but takes as an argument the name of the layout */
+static errno_t layout_load(const char *layout_name)
+{
+	/* TODO: change this into a loader for kb layout maps */
+	for (int i = 0; i < NUM_LAYOUTS; i++) {
+		if (str_cmp(layout[i]->name, layout_name) == 0) {
+			layout_active = layout[i];
+			layout_change(layout_active);
+			return EOK;
+		}
+	}
+
+	return ENOTSUP;
+}
+
+/* Handler for IPC call INPUT_CHANGE_LAYOUT */
+static void client_change_layout_handler(ipc_call_t *call)
+{
+	void *layout_name;
+	errno_t ret = async_data_write_accept(&layout_name, true, 0, 0, 0, 0);
+	if (ret != EOK) {
+		async_answer_0(call, ret);
+		return;
+	}
+
+	errno_t retval = layout_load((char *)layout_name);
+	free(layout_name);
+	async_answer_0(call, retval);
+}
+
+/* Handler for IPC call INPUT_GET_LAYOUT */
+static void client_get_layout_handler(ipc_call_t *call)
+{
+	const char *layout_name = layout_active->name;
+	size_t length = str_size(layout_name) + 1;
+	ipc_call_t id;
+
+	async_answer_1(call, EOK, length);
+	if (async_data_read_receive(&id, NULL)) {
```

Comment:
This is wrong. You can't reply to *call with EOK first. You first need to call async_data_read_receive() and if it fails, you need to reply with error to call.